executor/src/witgen/data_structures/copy_constraints.rs

```diff
@@ -0,0 +1,129 @@
+use std::collections::{BTreeMap, BTreeSet};
+
+/// Global copy constraints, represented as a graph on all cells. Each node has one outgoing edge
+/// that points to the next cell that should be equal to it, forming a set of cycles.
+pub struct CopyConstraints<CellId> {
+    /// The outgoing edge for each cell. If a cell is not in the map, it is its own next cell.
+    edges: BTreeMap<CellId, CellId>,
+}
+
+// Can't derive Default if CellId doesn't implement Default :o
+impl<CellId: Ord + Copy> Default for CopyConstraints<CellId> {
+    fn default() -> Self {
+        Self {
+            edges: Default::default(),
+        }
+    }
+}
+
+impl<CellId: Ord + Copy> CopyConstraints<CellId> {
+    pub fn new(constraint_pairs: &[(CellId, CellId)]) -> Self {
```

Comment:
Can you add a short doc string that explains if this just adds edges to the graph or if it forces the pairs to be in the same equivalence class.

Reply:
I mean this is a public function, it should be documented. What is the meaning of the CellId pairs? Does this need to be a full cross product of all items in an equivalence class, or what is it?

Reply:
Ah read the comment. Yes, that is what I meant :)